pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -61,23 +68,56 @@ func NewCache() *Cache {
 //
 // NodeCache objects are thread safe within the context of NodeCache,
 type NodeCache struct {
-	mu    sync.RWMutex
-	cache predicateMap
+	mu                 sync.RWMutex
+	cache              predicateMap
+	generation         uint64
+	snapshotGeneration uint64
+	// predicateGenerations stores generation numbers for predicates, incremented on
+	// predicate invalidation. Created on first update. Use 0 if does not
+	// exist.
+	predicateGenerations         map[string]uint64
+	snapshotPredicateGenerations map[string]uint64
 }
 
 // newNodeCache returns an empty NodeCache.
 func newNodeCache() *NodeCache {
 	return &NodeCache{
-		cache: make(predicateMap),
+		cache:                        make(predicateMap),
+		predicateGenerations:         make(map[string]uint64),
+		snapshotPredicateGenerations: make(map[string]uint64),
+	}
+}
+
+// Snapshot snapshots current generations of cache.
+func (c *Cache) Snapshot(nodes []*v1.Node) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for _, node := range nodes {
+		if _, exists := c.nodeToCache[node.Name]; !exists {
+			c.nodeToCache[node.Name] = newNodeCache()
```

Comment:
I thought earlier we must make sure all nodes has node caches because equivalence cache may not have caches for [nodes listed from lister](https://github.com/kubernetes/kubernetes/blob/v1.13.0-alpha.0/pkg/scheduler/core/generic_scheduler.go#L115) and `checkNode` will get a fresh node cache to use. Invalidations happens after snapshotting and before node cache being created will be ignored. (This is what `TestCacheInvalidationRace` tests)

But I find now it's not necessary because we run predicates on nodes in scheduler cache not nodes listed from lister after [this PR](https://github.com/kubernetes/kubernetes/pull/66733/files#diff-c237cdd9e4cb201118ca380732d7f361L367). When a node is added into scheduler cache, [it will create node cache for it](https://github.com/kubernetes/kubernetes/blob/e124159990c6d22e9ca662ef35de5eeaef5d8956/pkg/scheduler/factory/factory.go#L856).

I'll update this and move `GetNodeCache()` before `schedulerCache.AddNode` in `configFactory. addNodeToCache()` to make sure node cache is created before being adding into schedule cache.

Reply:
> I'll update this and move GetNodeCache() before schedulerCache.AddNode in configFactory. addNodeToCache() to make sure node cache is created before being adding into schedule cache.

This is what I'm thinking recently: we should try to avoid calling`GetNodeCache()` in any scheduling path as it holds a lock across nodes. You need to handle in node delete as well.

Reply:
@resouer Thanks! We should invalidate node before removing it from scheduler cache.